File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.Arrays;

class Solution {
    public long solution(int n, int[] times) {
        // 주어진 시간 동안 심사할 수 있는 인원을 기준으로 이분 탐색
        long left = 1;
        
        // 가능한 최대 시간(가장 오래걸리는 심사관의 소요 시간 * 모든 인원)으로 초기화
        long right = (long) Arrays.stream(times).max().getAsInt() * n;
        
        // 심사할 수 있는 인원의 초기 중앙 값은 최대값으로 설정
        long mid = right;
        
        long answer = 0;
        while(left <= right) {
            // 각 심사관이 주어진 시간(mid)동안 심사할 수 있는 최대 인원 계산
            long maxImmigrantsInMid = 0;
            for(int time : times) {
                maxImmigrantsInMid += (mid / time);
            }
            
            if(n <= maxImmigrantsInMid) {
                // mid시간 동안 심사가능한 인원이 n보다 큰 경우 -> 최소 시간을 찾아야한다.
                // 따라서 mid를 기준으로 왼쪽에 대해 이분 탐색 반복
                answer = mid;
                right = mid - 1;
            } else {
                // mid시간 동안 심사가능한 인원이 n보다 작은 경우 -> 모든 인원을 심사할 수 없으므로 시간 범위를 늘린다.
                left = mid + 1;
            }
            
            mid = left + (right - left) / 2;
        }
        
        return answer;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.kt
import java.util.Arrays

class Solution {
    fun solution(n: Int, times: IntArray): Long {
        // 맨 오른쪽은 (입력 값 중 가장 큰 값 * n)을 최댓값으로 지정
        var right: Long = Arrays.stream(times).max().asInt.toLong() * n
        var left: Long = 1

        var answer: Long = 0
        
        while(left <= right) {
            val mid: Long = left + (right - left) / 2
            
            // 각 심사관이 주어진 mid 시간 동안 심사할 수 있는 최대 인원 계산
            var maxImmigrantsInMid: Long = 0
            for(time in times) {
                maxImmigrantsInMid += (mid / time)
            }
            
            // n을 기준으로 이분탐색하여 n명을 심사할 수 있는 최소시간을 구한다.
            when {
                // 최대 심사가능한 인원이 n보다 크거나 같은 경우, 최소 시간을 구하기 위해 범위를 줄여서 반복한다.
                n <= maxImmigrantsInMid -> {
                    answer = mid
                    right = mid - 1
                }
                // 최대 심사 가능한 인원이 n보다 작은 경우 시간 범위를 늘린다.
                else -> left = mid + 1
            }
            
            // 중앙 값 갱신
        }
        return answer
    }
}